ENH: integrate: multidimensional integration of array-valued functions

ollybritton opened the pull request:
*For reference, this work is being completed as part of my internship at [Quansight Labs](https://labs.quansight.org/), which runs until the end of September.*

#### Reference issue
Addresses #20252.

#### What does this implement/fix?
This PR would enhance support for multidimensional integration ("cubature") through a new function `scipy.integrate.cubature`.

One of the current gaps in the capabilities of `scipy.integrate` is support for array-valued multidimensional integrals. At the moment, there is support for multidimensional *scalar-valued* integrands through `dblquad`, `tplquad` and `nquad`, and separate support for *1D* array-valued integrands in `quad_vec`, but not for simultaneously array-valued and multidimensional integrands.

Furthermore, `dblquad`, `tplquad` and `nquad` are implemented as iterated 1D integrals using `quad`, which means integration in a moderate number of dimensions can be slow. `quad` is implemented as a wrapper around the Fortran library QUADPACK, which makes it difficult to support other array APIs like `_tanhsinh` does.

These functions and `quad_vec` also do not support custom integration rules, which means it is hard to compute tricky multidimensional integrals with singularities.

This new `cubature` function has an interface similar to `quad_vec` but works towards resolving these issues by:

- Supporting array-valued multidimensional integrands (and scalar/1D integrals as a special case).
- Using an adaptive algorithm implemented in Python similar to `quad_vec`, with the eventual goal of supporting the array API. This works by subdividing the region being integrated until a tolerance is reached, rather than computing iterated 1D integrals.
- Allowing custom integration rules, which are classes with an `estimate(f, a, b)` method and an `estimate_error(f, a, b)` method. There is built-in support for Cartesian products of 1D rules like Newton-Cotes, Gauss-Legendre, Gauss-Kronrod, and support for "true cubature" rules like Genz-Malik which only work in $N>1$ dimensions, but also methods available to let users write their own rules.

#### Additional information
##### Interface details
<details>
<summary>Interface details (dropdown)</summary>

`cubature` is implemented in `integrate/_cubature.py` and has the signature:

```python
cubature(f, a, b,
	rule="gk21",
	rtol=1e-05, atol=1e-08,
	max_subdivisions=10000,
    args=(), kwargs=None)
```

`f` can be any callable which accepts arrays of shape `(npoints, ndim)` and outputs arrays of shape `(npoints, output_dim_1, ..., output_dim_2)`. `cubature` will then return an array of shape `(output_dim_1, ..., output_dim_n)` for the integral over the region specified by corners `a` and `b`, given as rank-1 arrays of length `ndim`.

`rule` is a string specifying the underlying cubature rule to use, and supports the same options as the `quadrature` option in `quad_vec`: `gk21`, `gk15` and `trapezoid`.

`rule` can also be an instance of a new class, `Rule`, which lives in `integrate/_rules/_base.py`. This is a general interface for numerical integration algorithms:

```python
class Rule:
	def estimate(self, f, a, b, args=(), kwargs=None): ...
	def estimate_error(self, f, a, b, args=(), kwargs=None): ...
```

There are then several subclasses. If a subclass does not implement `error_estimate`, then a default error estimator is used. The subclasses are:

- `FixedRule`, which is a rule implemented as the weighted sum of function evaluations at fixed nodes. For example, Gauss-Legendre quadrature.
- `NestedRule`, which is a rule where the error is estimated as the absolute difference between two underlying rules.
- `NestedFixedRule`, which is a rule where the error is estimated as the absolute difference between two underlying fixed rules.
- `ProductFixed`, which finds the `n`-dim cubature rule from the product of `n` 1D fixed rules
- `ProductNestedFixed`, which finds the `n`-dim cubature rule from the product of `n` 1D nested fixed rules

Implementations are then given in `integrate/_rules` for:

- `GaussKronrodQuadrature`, a `NestedFixedRule` implementing Gauss-Kronrod quadrature
- `GaussLegendreQuadrature`, a `FixedRule` implementing Gauss-Legendre quadrature
- `NewtonCotesQuadrature`, a `FixedRule` implementing Newton-Cotes (trapezoid) quadrature
- `GenzMalikCubature`, a `NestedFixedRule` implementing Genz-Malik cubature

This `Rule` class and its subclasses are currently not exposed as part of the public API.

</details>

##### Examples
(Examples taken from the docstring of `cubature`).

<details>
<summary>Vector-valued 1D integral (dropdown)</summary>

A simple 1D integral with vector output. Here $f(x) = x^n$ is integrated over the interval $[0, 1]$ for ten values of $n$. $f$ returns an array, so all ten integrals are done simultaneously. Since no rule is specified, the default `"gk21"` is used, which corresponds to `GaussKronrodQuadrature(21)`. Since the spatial dimension of the integrand is 1D, this could also be done using `quad_vec`. 

```python
>>> import numpy as np
>>> from scipy.integrate import cubature
>>> from scipy.integrate._rules import GaussKronrodQuadrature
>>> def f(x, n):
...    return x.reshape(-1, 1)**n  # Make sure x and n are broadcastable
>>> res = cubature(
...     f,
...     a=[0],
...     b=[1],
...     args=(
...         # Since f accepts arrays of shape (npoints, ndim) we need to
...         # make sure n is the right shape
...         np.arange(10).reshape(1, -1),
...     )
... )
>>> res.estimate
 array([1.        , 0.5       , 0.33333333, 0.25      , 0.2       ,
            0.16666667, 0.14285714, 0.125     , 0.11111111, 0.1       ])
```

</details>

<details>
<summary>Tensor-valued 7D integral (dropdown)</summary>

A 7D integral with arbitrary-shaped array output. Here $f(\pmb x) = \cos(2\pi r + \pmb \alpha^\top \pmb x)$ for some $r$ and $\pmb \alpha$, and the integral is performed over the unit hybercube, ``[0, 1]^7``.

```python
>>> import numpy as np
>>> from scipy.integrate import cubature
>>> from scipy.integrate._rules import GenzMalikCubature
>>> def f(x, r, alphas):
...     # f(x) = cos(2*pi*r + alpha @ x)
...     # Allow r and alphas to be arbitrary shape
...     npoints, ndim = x.shape[0], x.shape[-1]
...     alphas_reshaped = alphas[np.newaxis, :]
...     x_reshaped = x.reshape(npoints, *([1]*(len(alphas.shape) - 1)), ndim)
...     return np.cos(2*np.pi*r + np.sum(alphas_reshaped * x_reshaped, axis=-1))
>>> np.random.seed(1)
>>> res = cubature(
...     f=f,
...     a=np.array([0, 0, 0, 0, 0, 0, 0]),
...     b=np.array([1, 1, 1, 1, 1, 1, 1]),
...     rule=GenzMalikCubature(7),
...     kwargs={
...         "r": np.random.rand(2, 3),
...         "alphas": np.random.rand(2, 3, 7),
...     }
... )
>>> res.estimate
 array([[-0.61336595,  0.88388877, -0.57997549],
		[-0.86968418, -0.86877137, -0.64602074]])
```

</details>

##### Difficulties and next steps
It would be great to get some suggestions on the interface, especially around having `NestedRule` and `NestedFixedRule`.

Currently the following are not supported:

- Infinite endpoints
- Accepting a `points` parameter like `quad_vec`
- Broadcastable endpoints
- Non-rectangular regions

It is our hope that this will be added before the end of the internship (end of September), depending on how much further work is required for this PR.

A future PR can also update `quad_vec` to accept the 1D rule interface proposed here, perhaps just by using its existing "quadrature" argument. This would allow for plug-and-play rules in adaptive quadrature and would work nicely with integration methods such as tanh-sinh.

Many thanks to @izaid and @ev-br, who have given lots of valuable support for this PR.

ollybritton committed:
intermediate work on implementing cubature

ollybritton committed:
improve testing + minor changes to cub

ollybritton committed:
refactor testing

ollybritton committed:
refactor, add more tests, use `a` and `b` instead

ollybritton committed:
switch to opposite axis ordering, always output full result

ollybritton committed:
test integrals using closed forms

ollybritton committed:
support for arbitrary-shaped array-valued functions

ollybritton committed:
add closed form for f_4

ollybritton committed:
test arbitrarily-shaped params for f_4

ollybritton committed:
refactoring and cleanup

ollybritton committed:
change error estimate and reorder

ollybritton committed:
restructure and fix error estimation

ollybritton committed:
partially working genz-malik cubature

ollybritton committed:
improve genz-malik cubature

ollybritton committed:
style changes and generate newton cotes weights

ollybritton committed:
address some comments from meeting

ollybritton committed:
add docstrings

ollybritton committed:
add more docstrings

ollybritton committed:
almost remove dependence on inner classes

ollybritton committed:
use .rule rather than .nodes and .weights

documentation is now outdated

ollybritton committed:
return nodes and weights together

ollybritton committed:
change param generation for genz_malik_1980_f_2

ollybritton committed:
improvements to ErrorFromDifference and GenzMalik

ollybritton committed:
more tests and api changes

ollybritton committed:
test fix and docs

ollybritton committed:
push latest changes before weekly meeting

ollybritton committed:
refactor to use subdir and change some names

ollybritton committed:
add example to cub

ollybritton committed:
fix wrong function name in example

ollybritton committed:
rename and update docs

ollybritton committed:
more work on docs

ollybritton committed:
rename error_estimate as estimate_error

ollybritton committed:
docs update and move spatial dim to last axis

ollybritton committed:
update docs to reflect latest api changes

ollybritton committed:
fix imports in doctests

ilayn commented:
Thanks @ollybritton this is a very needed and nice addition. Here are two immediate comments from my side; 

 1- Why `cub` and not `cubature`? It does not seem like a good abbreviation given its very specific function.
 2- I think the OOP parts are a bit too much abstracted. Not saying it is bad but typically there is a very nice rule that many people voiced over decades; _"if you have a class which has two methods one of which is `__init__` that should have been a function. "_. I think it also applies here. I think this is the most famous one :) https://www.youtube.com/watch?v=o9pEzgHorH0

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,310 @@
+    Adaptive cubature of multidimensional array-valued function.
+
+    Given an arbitrary integration rule, this function returns an estimate of the
+    integral to the requested tolerance over the region defined by the arrays ``a`` and
```
Here and elsewhere arguments should use single backticks
```suggestion
    integral to the requested tolerance over the region defined by the arrays `a` and
```

j-bowhay reviewed (commented):
Thanks this looks nice and a useful thing to add!

Some questions:

1. This appears to be heavily geared to h-adaptive integration. Does the design rule out p-adaptive integration if we decide we want to add it in the future? Likewise, what about MC/QMC (such as that offered by [cuba](https://feynarts.de/cuba/))
2. Benchmarking. It would be nice to know how the performance compares (accuracy, function evaluations etc) to other implementations such as https://github.com/stevengj/cubature . There is a python wrapper https://github.com/saullocastro/cubature so this should not be too hard.

lucascolley commented on scipy/integrate/_rules/_base.py:
```diff
@@ -0,0 +1,657 @@
+        corresponding weights. ``nodes`` should be of shape ``(num_nodes,)`` for 1D
+        cubature rules (quadratures) and more generally for N-D cubature rules, it
+        should be of shape ``(num_nodes, ndim)``. ``weights`` should be of shape
+        `(num_nodes,)`. The nodes and weights should be for integrals over `[-1, 1]^n`.
```
```suggestion
        ``(num_nodes,)``. The nodes and weights should be for integrals over ``[-1, 1]^n``.
```

or `:math:` like elsewhere

lucascolley commented on scipy/integrate/_rules/_base.py:
```diff
@@ -0,0 +1,657 @@
+
+
+class NestedRule(Rule):
+    r"""
+    A rule with error estimate given by the difference between two underlying rules. If
+    constructed as ``NestedRule(higher, lower)``, this will use::
```
```suggestion
    r"""
    A rule with error estimate given by the difference between two underlying rules.
    
    If constructed as ``NestedRule(higher, lower)``, this will use::
```

and a few others

lucascolley commented on scipy/integrate/_rules/_base.py:
```diff
@@ -0,0 +1,657 @@
+        nodes, weights = self.nodes_and_weights
+        lower_nodes, lower_weights = self.lower_nodes_and_weights
+
+        error_nodes = np.concat([nodes, lower_nodes], axis=0)
+        error_weights = np.concat([weights, -lower_weights], axis=0)
```
we'll have to use `np.concatenate` for now to support NumPy<2.0 (until array API standard support is added)

lucascolley commented on scipy/integrate/_rules/_base.py:
```diff
@@ -0,0 +1,657 @@
+            [rule.nodes_and_weights[0] for rule in self.base_rules]
+        )
+
+        weights = np.prod(_cartesian_product(
+            [rule.nodes_and_weights[1] for rule in self.base_rules]
+        ), axis=-1)
```
this could definitely be clearer - a suggestion
```suggestion
        weights = np.prod(
            _cartesian_product(
                [rule.nodes_and_weights[1] for rule in self.base_rules]
            ),
            axis=-1,
        )
```

or perhaps clearer to put the `_cartesian_product([rule.nodes_and_weights[1] for rule in self.base_rules])` on a separate line

lucascolley reviewed (commented):
nice stuff @ollybritton! some comments from a look over everything apart from the tests

j-bowhay commented:
> Thanks this looks nice and a useful thing to add!
> 
> Some questions:
> 
> 1. This appears to be heavily geared to h-adaptive integration. Does the design rule out p-adaptive integration if we decide we want to add it in the future? Likewise, what about MC/QMC (such as that offered by [cuba](https://feynarts.de/cuba/))
> 2. Benchmarking. It would be nice to know how the performance compares (accuracy, function evaluations etc) to other implementations such as https://github.com/stevengj/cubature . There is a python wrapper https://github.com/saullocastro/cubature so this should not be too hard. Also, it would be good to compare the proposed rules, if one is strictly worse maybe it should not be included.

ollybritton commented:
Thanks for the feedback, everyone!

@ilayn, 1: Definitely open to suggestions for better names, both for `cub` and for everything else. The only motivation for `cub` rather than `cubature` was to match `quad` and `quad_vec`. It seems most people think `cubature` would be better, so I'm happy to change it.

2: I agree that this does go quite OOP-heavy, especially with the classes like `FixedRule`, `NestedRule` and `NestedFixedRule` – this isn't really a definite choice but rather because of a lack of immediate better alternatives. I'm open to suggestions about how this could be simplified or made less abstract – the details around the `Rule` classe and its subclasses is the part of the interface that I am least happy with.

`quad_vec` currently takes the approach of just using functions with signature `rule(a, b, f) -> est, err, round_err`. I think the problem with just using functions like this is that there needs to be some way to access the nodes being used so that you can take arbitrary products of 1D quadratures. And the reason for splitting up the `estimate` and `error_estimate` is to allow for error estimation that doesn't depend on the specific rule being used.

@j-bowhay, 1: Yes, at the moment this only supports $h$-adaptive integration but it would be nice to support also $p$-adaptive integration and Monte Carlo methods. This is not ruled out by this design, it could be added in the future by extending the `Rule` interface. `cub` could then improve the estimate over a region/axis with high error by increasing the number of nodes/sample points rather than by subdividing.

One thing that might be difficult is supporting $h$-adaptive and $p$-adaptive integration simultaneously, e.g. with an integral like

$$
\int_{0}^{1}\int_{-1}^{1}\frac{y}{\sqrt{1-x^{2}}}\text dx\text dy
$$

it would be nice to be able to use a rule like tanh-sinh to handle the endpoint singularities for the $x$-axis, but potentially a different rule for the the $y$-axis. Then if you were to increase the number of nodes in tanh-sinh rather than subdividing the $x$-axis, I think you would have to also re-evaluate the function across the rest of the domain (although maybe this is fine?). Definitely welcome to suggestions about how this could work, or comments about whether supporting examples like this would be necessary in the first place.

2: I'm happy to do some benchmarks.


@lucascolley: Thank you for the suggestions and feedback, I also see some of the CI is failing so I'll address that and your comments today.

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,310 @@
+    rtol: float
+
+
+def cub(f, a, b, rule="gk21", rtol=1e-05, atol=1e-08, max_subdivisions=10000,
+        args=(), kwargs=None):
```
I guess my comment about h-adaptive vs p-adaptive was sparked by this interface. `max_subdivisions` is only relevant for h-adaptive methods. I wonder if the signature should only contain arguments that are relevant for all rules and then rule specific options should be passed in a dict. This is how the interface for `minimize` works for example.

izaid replied:
I think this is a good comment. I'm not going to suggest a solution, but here are ways it go. I think if we were to restrict to arguments valid to _all_ rules then we'd only have `f`, `a`, and `b`, not even `rtol` and `atol`. If we were to restrict to arguments valid to adaptive, either p-adaptive or h-adaptive, then we'd keep the tolerances as well. But then maybe in this case we'd just end up with two functions, `p_cubature` and`h_cubature`, with `h_cubature` being what is implemented here. Both could make use of the same rule interface.

izaid replied:
And you might still want `cubature` to alias to one of them, perhaps `h_cubature`.

j-bowhay replied:
Agreed a few different ways to tackle this! Maybe wait and see what other people think

ollybritton committed:
DOC: address docs and style issues

ollybritton committed:
BUG: use np.concatenate instead of np.concat

ollybritton committed:
API: rename `cub` to `cubature`

ollybritton committed:
API: expand "quad" and "cub" suffixes for Rules

ollybritton committed:
API: allow specifying genz-malik as string rule

This means that you don't need to import the GenzMalikCubature class and
can instead specify it as "genz-malik" in the argument to `cubature`. The
examples in the docstring of `cubature` have also been updated to
reflect this, and so have the tests.

ollybritton committed:
MAINT: minor changes to _apply_fixed_rule

ollybritton committed:
intermediate work on parallelisation

ollybritton committed:
ENH: support `executor` argument and benchmarks

This adds an optional `executor` argument to `cubature` so that
subregions can be processed in parallel, and adds benchmarks to see how
the execution time compares.

ollybritton committed:
TST: mark certain tests as slow

ollybritton commented:
@j-bowhay I've run some benchmarks, both [informally in a notebook](https://gist.github.com/ollybritton/ef0fe44975ab33dc6abd8e6511c38f6f) and have also added some `asv` benchmarks to `benchmarks/integrate.py`. The results are mixed, for simple integrals that don't require any subdivision, `cubature` is 2-3x slower than `stevengj/cubature` using the Python wrapper.

For more complicated integrals, `cubature` can sometimes be 100x slower than `stevengj/cubature` which isn't ideal. In some high-dimensional situations `cubature` can be more competitive, ranging anywhere from 5x slower to ~2x faster (although these cases are rare). It's also difficult to compare `stevengj/cubature` to this new implementation using array-valued integrands since `stevengj/cubature` only supports 1D arrays (unless I am mistaken).

Since `stevengj/cubature` is GPLv3, I don't want to read the source because of licensing concerns. From profiling, I think the main reason that it can be 10x-100x faster is that it gets away with much fewer function evaluations than this `cubature` does – there's an example in the notebook above where it uses 30x more evaluations than `stevengj/cubature` for the same tolerances, so if the error estimation were improved it could be only 2-3x slower.

izaid commented:
> @j-bowhay I've run some benchmarks, both [informally in a notebook](https://gist.github.com/ollybritton/ef0fe44975ab33dc6abd8e6511c38f6f) and have also added some `asv` benchmarks to `benchmarks/integrate.py`. The results are mixed, for simple integrals that don't require any subdivision, `cubature` is 2-3x slower than `stevengj/cubature` using the Python wrapper.
> 
> For more complicated integrals, `cubature` can sometimes be 100x slower than `stevengj/cubature` which isn't ideal. In some high-dimensional situations `cubature` can be more competitive, ranging anywhere from 5x slower to ~2x faster (although these cases are rare). It's also difficult to compare `stevengj/cubature` to this new implementation using array-valued integrands since `stevengj/cubature` only supports 1D arrays (unless I am mistaken).
> 
> Since `stevengj/cubature` is GPLv3, I don't want to read the source because of licensing concerns. From profiling, I think the main reason that it can be 10x-100x faster is that it gets away with much fewer function evaluations than this `cubature` does – there's an example in the notebook above where it uses 30x more evaluations than `stevengj/cubature` for the same tolerances, so if the error estimation were improved it could be only 2-3x slower.

I just want to add on a bit to this - and @ollybritton, please correct me if I get something wrong. We had a good look at benchmarking, and generally came to the following conclusion. It's either a question of getting the parallelization right (and we've now added basic support for that) or getting better error estimates in the rules. Regarding the latter, this is why we thought hard about a sensible rule API including the `estimate_error` method, so that these can be improved over time.

Essentially, the error estimates we have in there now might be a bit too strict. Error estimation for numerical integration is a bit of a dark art, and we'd like to get all the basic structures in place so that in the future we or others can optimize this to get the balance correct between accuracy and efficiency.

I think we are on track to be competitive with cubature once that is done, especially noting cubature is written in C with Cython bindings, whereas we've just used pure NumPy. And the work here is still fast enough as it is today.

ilayn commented:
If you have some identified parts of the code that are deemed to be blockingly slow, I can help with Cython/C translations.

j-bowhay commented:
Thanks @ollybritton for looking into this! I agree @izaid this looks quite encouraging given we are comparing compiled code and pure Python. I think there might be two further things that might be worth looking at:

1. Benchmarking the different rules you are proposing to add against each other. For example, it wouldn't surprise me if the `trapezoid` rule is strictly worse than the others for all but the simplest examples? In which case we probably wouldn't want to include it.
2. Function evaluation. Clearly, compiled code will be faster but is the number of function evaluations needed to achieve a given error competitive?

For both these questions it would be helpful to consider them for a range of tolerances. Ie I think following style of work-precision benchmarks for both runtime and function evaluations is helpful:
![image](https://github.com/user-attachments/assets/c5744b54-5d1e-4f41-9562-284391d04b2a)

Of course, if you think this is a distraction from more important points please feel free to put your effort elsewhere!

ollybritton commented:
> If you have some identified parts of the code that are deemed to be blockingly slow, I can help with Cython/C translations.

Thanks @ilayn, this sounds helpful but we'd rather stick to pure Python for now as the current bottleneck in performance is with the error estimation technique rather than the actual computation, and because it would be nice to eventually support the array API. If using Cython/C would offer a speed-up big enough to outweigh the costs in maintainability then some help with this would definitely be appreciated.

ilayn commented:
> and because it would be nice to eventually support the array API.

We discussed this a few times before. That's not really a blocker if it gets in front of performance.

> If using Cython/C would offer a speed-up big enough to outweigh the costs in maintainability. 

That's also no problem as we are converting everything c/cython. So please don't consider architectural problems just now at this stage.

ollybritton committed:
API: use `workers` instead of `executor` argument

ollybritton commented:
To summarise the discussion here so far:

- The new function was originally called `cub`, but this has now been renamed to `cubature`.
- Some discussion around the OOP-heavy parts of the currently private rules interface since there are currently several simple and related classes: `FixedRule`, `NestedRule`, `NestedFixedRule`.
- `cubature` is currently geared towards $h$-adaptive integration, but in the future it could also support $p$-adaptive integration by extending the rules interface.
- Some benchmarks:
	- Compared to [stevengj/cubature](https://github.com/stevengj/cubature) (written in C), this code can be anywhere from ~2x faster to ~100x slower depending on the specific example.
	- The performance bottleneck at the moment appears to be the number of function evaluations needed to reach a given tolerance, since the error estimates being used are stricter than those for [stevengj/cubature](https://github.com/stevengj/cubature).
	- For the same number of function evaluations, this is only 2x-3x slower. Because of the design of the rules API, improving this can be done relatively easily in the future by modifying the `estimate_error` method.
	- The performance gap could then hopefully be closed by tuning parallelization (for which there is now support via a `workers` argument to `cubature`) and by potential use of C/Cython bindings.
- At the suggestion of @j-bowhay, more should be done benchmarking comparing this to other numerical integrators with work-precision graphs. This can be done after addressing the other next steps:
	- Infinite endpoints
	- Non-rectangular regions and variable transformations
	- Accepting a `points` parameter to better handle singularities
	- Broadcastable endpoints

lucascolley commented on scipy/integrate/_rules/__init__.py:
```diff
@@ -0,0 +1,13 @@
+from ._base import (
+    Rule, FixedRule,
+    NestedFixedRule, NestedRule,
+    ProductFixed, ProductNestedFixed
```
nit

```suggestion
    ProductFixed, ProductNestedFixed,
```

ollybritton committed:
MAINT: remove unused `_MockExecutor`

lucascolley commented on scipy/integrate/_rules/_genz_malik.py:
```diff
@@ -0,0 +1,202 @@
+        # (ndim, npoints)
+        nodes = nodes.T
+
+        w_1 = (2**self.ndim) * (12824 - 9120 * self.ndim + (400 * self.ndim**2)) \
+            / 19683
```
Something like this is better (feel free to vary the whitespace to your preference)

```suggestion
        w_1 = (
            2**self.ndim * (12824 - 9120*self.ndim + (400 * self.ndim**2)) / 19683
        )
```

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -0,0 +1,924 @@
+    Rule, FixedRule, ProductNestedFixed,
+    NestedFixedRule, NestedRule,
+    NewtonCotesQuadrature, GaussLegendreQuadrature, GaussKronrodQuadrature,
+    GenzMalikCubature
```
nit

```suggestion
    GenzMalikCubature,
```

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -0,0 +1,924 @@
+        b,
+        rule,
+        max_subdivisions=10,
+        args=(np.arange(5),)
```
```suggestion
        args=(np.arange(5),),
```

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -0,0 +1,924 @@
+
+    res = cubature(f, a, b, rule, kwargs={
+        "r": r,
+        "alphas": alphas
```
```suggestion
        "alphas": alphas,
```

lucascolley commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -0,0 +1,924 @@
+        err_msg=f"estimate_error={res.error}, subdivisions={res.subdivisions}"
+    )
+
+    assert res.status == "converged", f"estimate_error={res.error}, subdivisions=\
+{res.subdivisions}, true_error={np.abs(res.estimate - exact(a, b, *args))}"
```
```suggestion
    err_msg = (f"estimate_error={res.error}, "
               f"subdivisions= {res.subdivisions}, "
               f"true_error={np.abs(res.estimate - exact(a, b, *args))}")
    assert res.status == "converged", err_msg
```

lucascolley reviewed (commented):
just a few comments but looks pretty much perfect to me from a style perspective now!

ollybritton committed:
STY: address style issues

ollybritton commented:
@lucascolley Thanks for the feedback! That should all be addressed now.

ev-br commented:
Looks like all review feedback was addressed. 
To keep the ball rolling, I think it makes sense to land this in about a week's time --- unless there are further comments of course. Or if someone needs more time to review?

lucascolley commented:
> I think it makes sense to land this in about a week's time --- unless there are further comments of course. Or if someone needs more time to review?

maybe worth giving a heads-up on [the forum thread](https://discuss.scientific-python.org/t/support-for-multidimensional-array-valued-integration-in-scipy-integrate/1337)?

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,333 @@
+    a = np.atleast_1d(a)
+    b = np.atleast_1d(b)
+
+    if a.ndim != 1 or b.ndim != 1:
+        raise ValueError("a and b should be 1D arrays")
```
Not tested

j-bowhay replied:
```suggestion
    if a.ndim != 1 or b.ndim != 1:
        raise ValueError("`a` and `b` should be 1D arrays")
```

j-bowhay commented on scipy/integrate/tests/test_cubature.py:
```diff
@@ -0,0 +1,927 @@
+)
+
+from scipy.integrate import cubature
+
```
The tests file is slightly difficult to read as it is unclear what is testing the public cubature function and what is testing the rules' functionality. Maybe you organise the tests into classes to make it easier to follow

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,333 @@
+    ...     alphas_reshaped = alphas[np.newaxis, :]
+    ...     x_reshaped = x.reshape(npoints, *([1]*(len(alphas.shape) - 1)), ndim)
+    ...     return np.cos(2*np.pi*r + np.sum(alphas_reshaped * x_reshaped, axis=-1))
+    >>> np.random.seed(1)
```
Should use `np.random.default_rng()` and then the doc build will automatically add a seed

j-bowhay reviewed (changes requested):
Thanks for your work on this so far!

Not a full review but somethings to consider

ollybritton committed:
API: remove `kwargs`, don't let `a`, `b` be floats

This commit addresses most of the suggestions in the PR including:

- Refer to endpoints as 1D, not rank-1
- Tighten up default `rtol` and `atol`
- Add `__all__` to make it clear what is public
- Remove unneccesary check for `NotImplementedError`
- Update error message for `a` and `b` not being 1D arrays
- Test an error is raised when `a` and `b` are not 1D arrays
- Remove duplicated `_subregion_coordinates` function
- Use `np.random.default_rng` in tests
- Organise test_cubature into classes
- Remove `_max_norm` function
- Remove `kwargs` as an argument to `cubature`
- Don't let `a` and `b` be passed as floats, they have to be 1D arrays

ollybritton committed:
DOC: update default tolerances

ollybritton commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,333 @@
+        supported.
+    rule : str, optional
+        Rule used to estimate the integral. If passing a string, the options are
+        "gauss-kronrod" (21 node), "newton-cotes" (5 node) or "genz-malik" (degree 7).
```
The main motivation for offering it is so that it supports the same values as `quad_vec`'s `quadrature` argument but if you don't think that's a concern, I'm happy to remove it.

ollybritton committed:
DOC: explain when different rules should be used

ollybritton committed:
DOC: fix leftover `cub` instead of `cubature` typo

ev-br commented:
@j-bowhay Your review is listed as _changes requested_, so you're looking for some more time to review this, correct? No pressure, am just trying to avoid this PR falling through cracks :-).

ev-br commented:
Nah, next week sounds great :-)

j-bowhay commented:
Ok a very basic benchmark, still missing comparison with `nquad` and Steven Johnson's code but all I have time for in 20 minutes today.

I am still -1 on including Newton Cotes rules until there is a better justification for doing so. In this example using the "trapezoid" performs by far the worst and the error estimate seems to fail?

![image](https://github.com/user-attachments/assets/06d298af-041c-4f8c-93da-19ea0f6f65d9)


<details open>
<summary>Details</summary>
<br>

```
from scipy.integrate import cubature
import numpy as np
import matplotlib.pyplot as plt

count = 0
def genz_malik_1980_f_1(x, r, alphas):
    r"""
    .. math:: f_1(\mathbf x) = \cos\left(2\pi r + \sum^n_{i = 1}\alpha_i x_i\right)

    .. code-block:: mathematica

        genzMalik1980f1[x_List, r_, alphas_List] := Cos[2*Pi*r + Total[x*alphas]]
    """
    global count
    count +=1

    npoints, ndim = x.shape[0], x.shape[-1]

    alphas_reshaped = alphas[np.newaxis, :]
    x_reshaped = x.reshape(npoints, *([1]*(len(alphas.shape) - 1)), ndim)

    return np.cos(2*np.pi*r + np.sum(alphas_reshaped * x_reshaped, axis=-1))


def genz_malik_1980_f_1_exact(a, b, r, alphas):
    ndim = len(a)
    a = a.reshape(*([1]*(len(alphas.shape) - 1)), ndim)
    b = b.reshape(*([1]*(len(alphas.shape) - 1)), ndim)

    return (-2)**ndim * 1/np.prod(alphas, axis=-1) \
        * np.cos(2*np.pi*r + np.sum(alphas * (a+b)/2, axis=-1)) \
        * np.prod(np.sin(alphas * (a-b)/2), axis=-1)


a = np.array([0, 0])
b = np.array([10, 10])
r = np.array([1/4])
alphas = np.array([9, 10])


gk21_err = []
gk21_count = []
gk15_err = []
gk15_count = []
trapezoid_err = []
trapezoid_count = []
genz_malik_err = []
genz_malik_count = []
for tol in np.logspace(-3, -12, num=10):
    res = cubature(genz_malik_1980_f_1, a=a, b=b, args=(r, alphas), rtol=tol, atol=tol)
    gk21_err.append(np.abs(res.estimate - genz_malik_1980_f_1_exact(a,b,r,alphas)))
    gk21_count.append(count)
    count = 0
    res = cubature(genz_malik_1980_f_1, a=a, b=b, args=(r, alphas), rtol=tol, atol=tol,
                   rule="gk15")
    gk15_err.append(np.abs(res.estimate - genz_malik_1980_f_1_exact(a,b,r,alphas)))
    gk15_count.append(count)
    count = 0
    res = cubature(genz_malik_1980_f_1, a=a, b=b, args=(r, alphas), rtol=tol, atol=tol,
                   rule="trapezoid")
    trapezoid_err.append(np.abs(res.estimate - genz_malik_1980_f_1_exact(a,b,r,alphas)))
    trapezoid_count.append(count)
    count = 0
    res = cubature(genz_malik_1980_f_1, a=a, b=b, args=(r, alphas), rtol=tol, atol=tol,
                   rule="genz-malik")
    genz_malik_err.append(np.abs(res.estimate - genz_malik_1980_f_1_exact(a,b,r,alphas)))
    genz_malik_count.append(count)


plt.loglog(gk21_err, gk21_count, "-.", label="cubature: gk21")
plt.loglog(gk15_err, gk15_count, "-o", label="cubature: gk15")
plt.loglog(trapezoid_err, trapezoid_count, "-v", label="cubature: trapezoid")
plt.loglog(genz_malik_err, genz_malik_count, "-p", label="cubature: genz_malik")
plt.xlabel("Function Evaluations")
plt.ylabel("Error")
plt.legend()
plt.show()
```

</details>

ollybritton commented:
@j-bowhay Thanks for the benchmarks ! I've added in a comparison against `stevengj/cubature` and `nquad` below.

![fig-full](https://github.com/user-attachments/assets/4620c192-8c97-439d-974c-65f023ecefbe)

I've also additionally measured the number of points at which `f` is evaluated at and the time it took on my machine. Since the new code and `stevengj/cubature` are both vectorized, `steveng/cubature` uses less calls `f(x)` but over the course of all the calls, the sum of `len(x)` is higher. `nquad` is not vectorized, so the number of function evaluations and points used are the same.

The graphs are a bit oddly shaped, but I think this is a consequence of approximate error estimation. Reducing the tolerance might have made the library choose a better interval for subdivision just by luck. For all the libraries, often the actual error is several orders of magnitude better than the requested tolerance.

Here is the modified code: <https://gist.github.com/ollybritton/24b6fb6b54217667064c2508d829b983>.

It does indeed fail to convege using `trapezoid` and for all but the weakest tolerance using `genz-malik` (`res.status == "not_converged"`).

I've also tried to run a similar benchmark for a 4-D integral over $[0, 10]^4$. The new `cubature` takes a long time but eventually returns when using `gk15`, `gk21` or `genz-malik`. I'm struggling to bench `steveng/cubature` since it is producing segfaults and `nquad` didn't return anything in time. Reducing the bounds to $[0, 1]^4$ gives the following results:

![fig_full_4d](https://github.com/user-attachments/assets/14f69177-b6c4-49b1-81bf-ec485112f54f)

(`steveng/cubature` was only tested for tolerances from $10^{-3}$ to $10^{-5}$ since otherwise it was using too much memory, and I have also ommited `trapezoid` here completely).

I agree about not including `trapezoid`, and even think that given `genz-malik`'s poor performance it might not be worth including that either. They could always remain private for now and then if the `Rule`s API were to be made public, people
could use them that way.

j-bowhay reviewed (commented):
Thanks for those benchmarks, they are very nice! I have also performed further benchmarks and was reaching similar conclusions:

The `gk15`, `gk21` rules perform very nicely and look like a great addition, +1 to include.

`genz-malik` doesn't seem to perform amazingly. I would be -0 on including as it doesn't seem to bring much to the table. If we find a use case where it performs better or if we can find another true cubature rule that performs better then I would be happy to see it included.

`trapezoid` (and newton cotes rule) don't seem to work well in practice. IMO these are best left to learn about in numerical analysis classes rather than practical use. -1 to include.

A few other minor comments from taking another look. I will be without reliable internet from next week onwards for a month so feel free to proceed without me @ev-br

j-bowhay commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,318 @@
+    rtol : float, optional
+        Relative tolerance. Default is 1e-08.
+    atol : float, optional
+        Absolute tolerance.
```
or feel free to combine together like https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html

ollybritton committed:
DOC: improve `rtol` and `atol` docs

- Document the default values
- Add an explanation of the stopping condition like `solve_ivp`

ollybritton committed:
BENCH: bench `cubature` for different `rtol`

Also add simple benchmarks for `nquad` and `quad_vec`.

ollybritton committed:
TST: remove unnecessary slow test

ollybritton committed:
API: remove Newton-Cotes as `cubature` option

ollybritton committed:
MAINT: remove unused `NestedRule` and `ProductNested`

ollybritton committed:
DOC: add further explanation to `CubatureResult`

ollybritton commented:
Thanks for the feedback! I've addressed this in the latest commits, added some more `asv` benchmarks and also improved some of the documentation. Should hopefully be almost ready now!

ev-br commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,316 @@
+            (npoints, ndim)
+
+        and output arrays of shape::
+            (npoints, output_dim_1, ..., output_dim_n)
```
I suspect `::` needs a blank line to render properly, here and elsewhere:

```suggestion

            (npoints, output_dim_1, ..., output_dim_n)
```

See https://output.circle-artifacts.com/output/job/3b105100-89ce-43d8-8cbb-b3c68f3daac7/artifacts/0/html/reference/generated/scipy.integrate.cubature.html#scipy.integrate.cubature

ev-br commented on scipy/integrate/_cubature.py:
```diff
@@ -0,0 +1,316 @@
+
+    Genz-Malik is typically less accurate than Gauss-Kronrod but has much fewer nodes,
+    so in this situation using "genz-malik" might be preferable.
+    """
```
Please duplicate the `References` section from private `_rules` objects so that they are user-visible,

ev-br commented on scipy/integrate/_rules/_gauss_kronrod.py:
```diff
@@ -0,0 +1,178 @@
+                -0.930157491355708226001207180059508,
+                -0.973906528517171720077964012084452,
+                -0.995657163025808080735527280689003,
+            ])
```
Please note the origin of the values.

ev-br reviewed (approved):
Several minor comments, all can be addressed in a follow-up.

Great job @ollybritton! Let's land this and keep the ball rolling. And thanks Jake, Lucas, Ilhan for the reviews.

lucascolley commented:
congrats 🎉 !

charliebritton commented:
Well done on having your code merged Olly I know you've been working hard all summer :)
